### PR TITLE
ENH: addition of carbon trackers to estimate compute costs

### PR DESCRIPTION
**Pull request by nikhil153**
## Changes proposed in this pull request
1. Addition of **carbontracker** feature to estimate carbon emissions incurred by processing workflows. 
2. Addition of [CodeCarbon](https://mlco2.github.io/codecarbon/index.html) dependency.  

This PR is in reference to #2481 issue. 

## Documentation that should be reviewed

All the code changes and additions are described in this [README](https://github.com/nikhil153/fmriprep/blob/carbon-trackers/carbontracker/README.md). In summary, following files are to be reviewed:

### Files modified:
1. [setup.cfg](https://github.com/nikhil153/fmriprep/blob/carbon-trackers/setup.cfg#L25) ([CodeCarbon](https://mlco2.github.io/codecarbon/index.html) dependency.) 
2. [parser.py](https://github.com/nikhil153/fmriprep/blob/carbon-trackers/fmriprep/cli/parser.py)
3. [run.py](https://github.com/nikhil153/fmriprep/blob/carbon-trackers/fmriprep/cli/run.py)
4. [config.py](https://github.com/nikhil153/fmriprep/blob/carbon-trackers/fmriprep/config.py)

### Files added 
1. [fmriprep/carbontracker](https://github.com/nikhil153/fmriprep/tree/carbon-trackers/carbontracker)
```
├── emissions_demo.csv
├── emissions_visual_demo.ipynb
├── figs
│   ├── CodeCarbon_example_emissions.png
│   ├── CodeCarbon_example_global.png
│   ├── fmriprep_carbontracker.jpg
│   └── sea-sig_logo.png
├── fmriprep_with_carbon_trackers.def
├── README.rst
├── requirements.txt
├── run_fmriprep_patch.sh
```

### Usage 
Carbon tracker feature can be used by specifying two optional arguments:
--track-carbon 
--country-code "CAN"

### CODE REVIEW

**Comment by nikhil153** [comment]
Hi @effigies - finally got a chance to wrap-up this ENH we talked about at the brainhack! Let me know if you think of any improvements for simplifying usage. Thanks!

**Comment by effigies** [comment]
@nikhil153 Thanks! I plan to look at this later this week for some initial feedback, but please bug me on Tuesday (holiday Monday) if I fail.

**Comment by nikhil153** [comment]
Hi @effigies - bugging you as asked :) Any updates?

**Comment by nikhil153** [comment]
Hi @effigies - I am seeing a couple of new conflicts on this thread. Any suggestions / comments?

**Comment by effigies** [comment]
Resolved conflicts. Had a quick read-through, and this seems reasonably constrained. We'll probably want to push the dependencies into an extra (they'll get installed in the Docker image, but they shouldn't be required to install the package). I'm running some unrelated tests, but I can try patching this in in the next couple days to get a sense of how it works.

**Comment by nikhil153** [comment]
Hi @effigies, sounds good, thanks so much!

**Comment by effigies** [comment]
Running with `fmriprep-docker -u $UID` (user mode instead of root), I see:

```
[codecarbon ERROR @ 16:30:00] Unable to read Intel RAPL files for CPU power, we will use a constant for your CPU power. Please view https://github.com/mlco2/codecarbon/issues/244 for workarounds : [Errno 13] Permission denied: '/sys/class/powercap/intel-
rapl/intel-rapl:1/energy_uj'                                                                                                                                                                                                                                  
[codecarbon ERROR @ 16:30:00] Unable to read Intel RAPL files for CPU power, we will use a constant for your CPU power. Please view https://github.com/mlco2/codecarbon/issues/244 for workarounds : [Errno 13] Permission denied: '/sys/class/powercap/intel-
rapl/intel-rapl:0/energy_uj'                                                                                                                                                                                                                                  
[codecarbon INFO @ 16:30:00] Tracking Intel CPU via RAPL interface                                                                                                                                                                                            
[codecarbon ERROR @ 16:30:01] Unable to read Intel RAPL files for CPU power, we will use a constant for your CPU power. Please view https://github.com/mlco2/codecarbon/issues/244 for workarounds : [Errno 13] Permission denied: '/sys/class/powercap/intel-
rapl/intel-rapl:1/energy_uj'                                                                                                                                                                                                                                  
[codecarbon ERROR @ 16:30:01] Unable to read Intel RAPL files for CPU power, we will use a constant for your CPU power. Please view https://github.com/mlco2/codecarbon/issues/244 for workarounds : [Errno 13] Permission denied: '/sys/class/powercap/intel-
rapl/intel-rapl:0/energy_uj'      
[codecarbon INFO @ 16:30:01] >>> Tracker's metadata:                                                                                                                                                                                                          
[codecarbon INFO @ 16:30:01]   Platform system: Linux-5.15.0-43-generic-x86_64-with-glibc2.31                                                                                                                                                                 
[codecarbon INFO @ 16:30:01]   Python version: 3.9.12                                                                                                                                                                                                         
[codecarbon INFO @ 16:30:01]   Available RAM : 62.544 GB                                                                                                                                                                                                      
[codecarbon INFO @ 16:30:01]   CPU count: 20                                                                                                                                                                                                                  
[codecarbon INFO @ 16:30:01]   CPU model: Intel(R) Core(TM) i9-10900 CPU @ 2.80GHz                                                                                                                                                                            
[codecarbon INFO @ 16:30:01]   GPU count: None                                                                                                                                                                                                                
[codecarbon INFO @ 16:30:01]   GPU model: None 
...
[codecarbon INFO @ 16:45:03] Energy consumed for RAM : 0.004693 kWh. RAM Power : 23.4539794921875 W
[codecarbon INFO @ 16:45:03] Energy consumed for all CPUs : 0.000000 kWh. All CPUs Power : 0.0 W
[codecarbon INFO @ 16:45:03] 0.004693 kWh of electricity used since the begining.
```

Is a 0W fallback expected?

**Comment by nikhil153** [comment]
So Docker version won't work unless run with the `--privileged` flag since by default it doesn't have sufficient permissions to read RAPL logs. I have been primarily using and testing with Singularity, which is more common on HPCs and that's where primarily we want to track carbon emissions.

File: fmriprep/cli/run.py [hunk 1]
```diff
@@ -39,6 +39,19 @@ def main():
...
+    if config.execution.track_carbon:
+        from codecarbon import OfflineEmissionsTracker
+
+        country_iso_code = config.execution.country_code
+        CC_log_dir = "/output/"
+        config.loggers.workflow.log(25, "CodeCarbon tracker started ...")
+        config.loggers.workflow.log(25, f"Using country_iso_code: {country_iso_code}")
+        config.loggers.workflow.log(25, f"Saving logs at: {CC_log_dir}")
+
+        tracker = OfflineEmissionsTracker(output_dir=CC_log_dir, country_iso_code=country_iso_code)
```
Comment (effigies) [inline fmriprep/cli/run.py:52]: I got:

```console
[codecarbon WARNING @ 19:45:41] graceful shutdown. Exceptions:                                                                                                                                                                                                
[codecarbon WARNING @ 19:45:41] <class 'Exception'>                                                                                                                                                                                                           
Traceback (most recent call last):                                                                                                                                                                                                                            
  File "/opt/conda/lib/python3.9/site-packages/codecarbon/core/util.py", line 18, in suppress                                                                                                                                                                 
    yield
  File "/opt/conda/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/opt/conda/lib/python3.9/site-packages/codecarbon/emissions_tracker.py", line 430, in stop
    persistence.out(emissions_data)
  File "/opt/conda/lib/python3.9/site-packages/codecarbon/output.py", line 143, in out
    df.to_csv(self.save_file_path, index=False)
  File "/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py", line 3551, in to_csv
    return DataFrameRenderer(formatter).to_csv(
  File "/opt/conda/lib/python3.9/site-packages/pandas/io/formats/format.py", line 1180, in to_csv
    csv_formatter.save()
  File "/opt/conda/lib/python3.9/site-packages/pandas/io/formats/csvs.py", line 241, in save
    with get_handle(
  File "/opt/conda/lib/python3.9/site-packages/pandas/io/common.py", line 697, in get_handle
    check_parent_directory(str(handle))
  File "/opt/conda/lib/python3.9/site-packages/pandas/io/common.py", line 571, in check_parent_directory
    raise OSError(rf"Cannot save file into a non-existent directory: '{parent}'")
OSError: Cannot save file into a non-existent directory: '/output'
```

Let's use the log directory:

```suggestion
        config.loggers.workflow.log(25, "CodeCarbon tracker started ...")
        config.loggers.workflow.log(25, f"Using country_iso_code: {country_iso_code}")
        config.loggers.workflow.log(25, f"Saving logs at: {config.execution.log_dir}")

        tracker = OfflineEmissionsTracker(output_dir=config.execution.log_dir, country_iso_code=country_iso_code)
```

File: fmriprep/cli/run.py [hunk 2]
```diff
@@ -179,6 +192,13 @@ def main():
     finally:
         from pkg_resources import resource_filename as pkgrf
 
+        # Code Carbon
+        if config.execution.track_carbon:
+            emissions: float = tracker.stop()
+            config.loggers.workflow.log(25, "CodeCarbon tracker has stopped.")
+            config.loggers.workflow.log(25, f"Saving logs at: {CC_log_dir}")
```
Comment (effigies) [inline fmriprep/cli/run.py:199]:
```suggestion
            config.loggers.workflow.log(25, f"Saving logs at: {config.execution.log_dir}")
```

**Comment by nikhil153** [comment]
Thanks @effigies! 
The `script` directory has a couple my test CSVs - which can be ignored. The other two scripts `[fetch_templates.py](https://github.com/nikhil153/fmriprep/blob/carbon-trackers/scripts/fetch_templates.py)` and `[generate_reference_mask.py](https://github.com/nikhil153/fmriprep/blob/carbon-trackers/scripts/generate_reference_mask.py)` are not mine. 
The `carbontracker` directory can also be ignored as it basically contains READMEs and scripts to patch my fork on the fMRIPrep Singularity container for trials and tests. 
What's the best way to migrate my existing documentation for this ENH to the official one?

**Comment by effigies** [comment]
I would probably go ahead and make a new top-level document, like `docs/carbon.rst`, which can be inserted into the TOC here:

https://github.com/nipreps/fmriprep/blob/53765b58e5924547feeea759552d1c147b9e4db9/docs/index.rst#L12-L22

I would place it above `faq`. Are you comfortable writing RST? If you'd rather stick to markdown, we would need to add `myst_parser` to the extensions and doc requirements:

https://github.com/nipreps/fmriprep/blob/53765b58e5924547feeea759552d1c147b9e4db9/docs/conf.py#L37-L47

https://github.com/nipreps/fmriprep/blob/53765b58e5924547feeea759552d1c147b9e4db9/setup.cfg#L54-L59

**Comment by effigies** [comment]
@nikhil153 I merged the condensed result: https://github.com/nipreps/fmriprep/commit/b6c8e1bfd2dd153663ac6bc17ec905ddf1e72b11

Documentation can be updated at your convenience.

**Comment by nikhil153** [comment]
@effigies Thanks so much! Really appreciate your help in this project! Curious to see carbon estimate across hardware and geolocations. 
RST is not a problem. I will update the documentation within few weeks!

**Pull request merged**